Add the Memory.Initialization module to the standard modules

dlongnecke-cray opened the pull request:
Add the Memory.Initialization module to the standard modules (#17118)

This PR resolves #16172 by adding a submodule to `Memory` called
`Initialization`. The `Initialization` module provides functions to
support low-level moves and deinits.

The intention behind providing such an unsafe module is to support
the same patterns used to implement standard collections in a
user facing way.

The `needsDeinit()` and `explicitDeinit()` functions can be used to
explicitly deinitialize a value. The compiler is not currently
aware of `explicitDeinit()`, so it should only be used on a slot
of an unmanaged buffer. Calling `explicitDeinit()` on a stack
allocated value that will go out of scope may result in the value
being deinitialized twice.

Various combinations of the `move...` family of functions can be used
to manipulate collection memory.

The `moveInitialize()` function can be used to move a distinct value
into the memory referred to by a reference. The `in` intent on its
second argument `rhs` ensures that `rhs` will be copy-initialized or
elided before it is moved into place.

The `moveToValue()` function can be used to remove a value from a
collection buffer (e.g., when used to implement a function such as
`list.pop()` which can return a value).

A combination of the above two functions can be used to create a
completely general purpose "move" function.

Future work:

- Make the compiler aware of `explicitDeinit` calls on variables,
  e.g. by removing `FLAG_AUTO_DESTROY`
- Use the bulk transfer interface for `moveInitializeArrayElements`
- Have the compiler emit an error when a copy occurs for the
  `in` formal `src` of `moveInitialize`

TESTING:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] Docs hosted for review

Reviewed by @vasslitvinov and @mppf. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

dlongnecke-cray commented:
Link to documentation is here: http://web.us.cray.com/~dlongnecke/add-memory-initialization-module/modules/standard/Memory/Initialization.html

dlongnecke-cray commented:
@vasslitvinov Would you be willing to be the primary reviewer for this?

@mppf Would you be willing to double check the `moveInitializeArrayElements` functions?

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+    :arg t: A type to check for a deinitializer
+    :type t: `type`
+
```
I think all types have to "support" deinitialization, it's just for some of them it is a no-op. Perhaps in chpldoc call it "if a given type has a non-empty deinitalizer" ?

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+    :return: A reference moved into a new value
+    :rtype: `t`
+  */
```
I think this should be called `moveToReturn`.
Things can be moved into "variables", not into "values".
I can move something into an integer, for example,
only because here "integer" is a slang for an "integer variable". :)

Not sure how to write the chpldoc... perhaps:

    Give ownership of the referenced value to the caller.

We cannot say "pass" the ownership, as it is not really taken away from the argument.

:arg arg: the value to consume [the reader will see that it is passed by const-ref]
:return: the value referenced by the argument

dlongnecke-cray replied:
I think the line between variable and value here is kind of thin, because what if I do something like...

```chapel
// What about this?
moveToValue(foobar);

// As opposed to this?
var x = moveToValue(foobar);
```

Here the function `moveToValue` is returning a variable moved into a new value, which may or may not be assigned to a `var`.

The name `moveToReturn` doesn't make much sense to me because it describes what the function is doing, not what it is used for. That we move some memory into a temporary variable within the function and then return it does not matter, because once returned it may or may not be captured into a variable.

dlongnecke-cray replied:
We should probably move further discussion to #16172 though.

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+  /*
+    Swap the values referred to by two references. No assignment takes
+    place when swapping ``lhs`` and ``rhs``.
```
... referred to by ``lhs`` and ``rhs``. By contrast with the operator `<=>`,
`moveSwap` does not perform assignments and initializations.

mppf replied:
I think the comment needs to make it clear that this function also does not call `<=>`.

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+  /*
+    Move-initialize a group of array elements from a different group of
+    elements in the same array.
```
Perhaps add: "This is equivalent to and potentially better-performing than a sequence of individual calls to `moveInitialize().`" Now looking at the code, perhaps "better performing" is better left out.

You need to define what happens when the two sets of indices overlap and present the error conditions that you check for.

dlongnecke-cray replied:
The reason I omitted those is because it ended up being a ridiculous amount of warnings. Is there a more eloquent way to put it?

vasslitvinov replied:
Possibly:
"This function reports an error when [this normal piece of code] does."
or "Calling this function is legal only when [some piece of code] is."

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+    Move-initialize a group of array elements from a different group of
+    elements in the same array.
+
+    :arg a: An array to move-initialize
```
How about: The array with the source and destination elements

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+  /*
+    Move-initialize a group of array elements from a different group of
+    elements in another array.
```
Remove "different"
Add, as above: This is equivalent to and potentially better-performing than a sequence of individual calls to `moveInitialize().`

vasslitvinov reviewed (approved):
Sounds good!
Please follow up on my feedback.

mppf commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+  }
+
+  /*
+    Check to see if a given type supports deinitialization.
```
> if a given type has a non-empty deinitalizer

That leaves a question about `proc MyRecord.deinit() { }` because the compiler adds calls to field deinits in that case. How about "Check to see if a given type needs to be deinitialized".

mppf commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+      Chapel is not aware of a call to ``explicitDeinit()``, which means
+      that the deinitializer of ``arg`` may run again once it goes out of
+      scope.
```
In Vass' suggestion - called only deinitialization -> called only when deinitialization - otherwise I am happy with it.

mppf commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+
+  /*
+    Move-initialize a reference from a value. No assignment is performed
+    when moving ``rhs`` into the memory referred to by ``lhs``.
```
This function can run the copy initializer today, for example if the actual argument for `rhs` is a reference. We should make that case into a compilation error but it is OK to leave that as future work.

mppf replied:
(Just be sure to make a TODO issue that we need to do that follow up task).

vasslitvinov replied:
>This function can run the copy initializer today

Uh-huh, then it is not a `move` and we need to do more explaining.

Maybe present it as equivalent to `proc init=(in rhs)` except (a) it can be invoked explicitly and so (b) overwrites whatever was in `lhs` before ?

mppf replied:
Oh, I don't think we should present it that way. I view it as a bug that it can run the copy initializer today. We could put a `.. note::` section about it if you want, but I hope David just makes a future work issue and we attend to it soon.

dlongnecke-cray replied:
I am hesitant to use the word `ownership` here. To me it implies `owned` classes. How about something like...

```
    Move-initialize ``lhs`` with the value in ``rhs``. The contents of ``rhs``
    are not deinitialized before the move, and ``rhs`` is not deinitialized
    after the move.
```

mppf commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+  }
+
+  private inline proc _errorNonCopyableType(type t) {
+    if !isCopyableType(t) then
```
Why wouldn't you be able to move initialize something that is not copyable?

dlongnecke-cray replied:
I guess because the alternative is saying that you're stealing elements from one array and putting them in another. Is that what this operation is doing? Just not how I visualized it.

dlongnecke-cray replied:
From reading all the other comments I'm assuming the answer is yes, it should always steal from `srcA`. So in that case should I just use a `_move` instead of `moveInitialize`?

mppf replied:
Sure

mppf commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,290 @@
+      ref dst = a[dstIdx];
+      ref src = a[srcIdx];
+      moveInitialize(dst, src);
+    }
```
Can you make an issue for the future work of optimizing communication for this and the other moveInitializeArrayElements function?

In particular, I am expecting that they use the bulk transfer interface which is what happens in the similar code in the Sort module: https://github.com/chapel-lang/chapel/blob/master/modules/packages/Sort.chpl#L1889-L1958 .

dlongnecke-cray replied:
Oh, there totally could be communication if the array is `dmapped`, huh...

dlongnecke-cray commented:
Hi @vasslitvinov I believe I addressed most of your feedback. I made some slight tweaks: in particular I avoid using the word `ownership`. I also left `moveToValue` as the same, we can debate it in #16172 more if you have a strong stance about `moveToReturn` being a better name.

I believe I am using a consistent language now consisting of words like "variable", "value", "contents of". No more use of "memory".

I made 3 work items for future TODOs.

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,302 @@
+      This function will halt if the value of ``numElements`` is negative
+      or is greater than `a.size`. It will also halt if any indices in
+      the range `dstStartIndex..(dstStartIndex + numElements)` or
+      `srcStartIndex..(srcStartIndex + numElements)` are out of bounds.
```
This function will halt if the value of ``numElements`` is negative
or any of the indices in `dstStartIndex..#numElements` or
`srcStartIndex..#numElements` are out of bounds.

dlongnecke-cray replied:
Are you suggesting to omit the paren expression or just to make use of the `#` operator?

vasslitvinov replied:
Right, my edit contains several suggestions:
* make the ranges correct (I think your formulas grab an extra index)
* switch to using the # operator as the more chapel-rific way to specify an n-element range
* dropping the paren expression is a side effect of the previous change
* drop "is greater than a.size" - because this is covered by the other conditions

vasslitvinov replied:
I wonder whether we should make it clear that there will be no halting upon `--fast` and `--no-checks`.

vasslitvinov commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,302 @@
+      `srcStartIndex..(srcStartIndex + numElements)` are out of bounds.
+
+    :arg a: The array with source and destination elements
+    :type a: `[]`
```
I think the :type a: line does not add value, consider removing it.

dlongnecke-cray replied:
I switched from the `:type b: t` to `:type b: The type of a` throughout. Thanks!

vasslitvinov reviewed (approved):
Looks good. Please see my comments.

dlongnecke-cray commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,302 @@
+  */
+  proc moveSwap(ref lhs: ?t, ref rhs: t) {
+    if t == nothing {
+      return none;
```
Whoops, that's definitely a bug. Thanks!

dlongnecke-cray commented on modules/standard/Memory/Initialization.chpl:
```diff
@@ -0,0 +1,302 @@
+  /*
+    Move-initialize a group of array elements from a group of elements in the
+    same array. This function is equivalent to a sequence of individual calls
+    to :proc:`moveInitialize()`.
```
I think the behavior is fairly well defined, though, isn't it? It's just the outcome is bad, but even if the ranges don't overlap you're going to have a bunch of uninitialized elements in your array.

vasslitvinov replied:
Maybe we need to say what part of the behavior is undefined. Specifically, if A[i] is move-initialized from A[j] and A[j] is move-initialized from A[k], it is undefined whether A[i] will be move-initialized with A[j]'s new value or old value. Also, the answer may be different for different indices and vary from run to run. 

I am not proposing to write this whole thing, however we should indicate that this is undefined.

dlongnecke-cray replied:
I totally forgot that it's happening in parallel! Do you think I should just add a check to halt if the ranges overlap?

vasslitvinov replied:
I would, except for a performance concern. Maybe, like in the other case, do the overlap check with `if boundsChecking` ?
